### ISSUE TITLE
0.29.0: test suite is failing

### ISSUE THREAD

**Issue by kloczek** [commonmark/cmark#350]
```
+ /usr/bin/make -O -j48 V=1 VERBOSE=1 -C x86_64-redhat-linux-gnu test
make: Entering directory '/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu'
Running tests...
/usr/bin/ctest --force-new-ctest-process
Test project /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu
    Start 1: api_test
1/9 Test #1: api_test .........................   Passed    0.01 sec
    Start 2: html_normalization
2/9 Test #2: html_normalization ...............***Failed    0.20 sec
    Start 3: spectest_library
3/9 Test #3: spectest_library .................   Passed    0.15 sec
    Start 4: pathological_tests_library
4/9 Test #4: pathological_tests_library .......   Passed    2.78 sec
    Start 5: roundtriptest_library
5/9 Test #5: roundtriptest_library ............***Failed    0.16 sec
    Start 6: entity_library
6/9 Test #6: entity_library ...................   Passed    0.11 sec
    Start 7: spectest_executable
7/9 Test #7: spectest_executable ..............   Passed    3.28 sec
    Start 8: smartpuncttest_executable
8/9 Test #8: smartpuncttest_executable ........   Passed    0.24 sec
    Start 9: regressiontest_executable
9/9 Test #9: regressiontest_executable ........   Passed    0.22 sec

78% tests passed, 2 tests failed out of 9

Total Test time (real) =   7.16 sec

The following tests FAILED:
          2 - html_normalization (Failed)
          5 - roundtriptest_library (Failed)
Errors while running CTest
make: *** [Makefile:153: test] Error 8
make: Leaving directory '/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu'
```
In Testing/Temporary/LastTest.log I found:
```
[tkloczko@barrel Temporary]$ cat LastTest.log
Start testing: Jul 18 11:15 BST
----------------------------------------------------------
1/9 Testing: api_test
1/9 Test: api_test
Command: "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/api_test/api_test"
Directory: /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir
"api_test" start time: Jul 18 11:15 BST
Output:
----------------------------------------------------------
538 tests passed, 0 failed, 0 skipped
PASS
<end of output>
Test time =   0.01 sec
----------------------------------------------------------
Test Passed.
"api_test" end time: Jul 18 11:15 BST
"api_test" time elapsed: 00:00:00
----------------------------------------------------------

2/9 Testing: html_normalization
2/9 Test: html_normalization
Command: "/usr/bin/python3" "-m" "doctest" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/normalize.py"
Directory: /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir
"html_normalization" start time: Jul 18 11:15 BST
Output:
----------------------------------------------------------
**********************************************************************
File "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/normalize.py", line 169, in normalize.normalize_html
Failed example:
    normalize_html('<a title="bar" HREF="foo">x</a>')
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib64/python3.8/doctest.py", line 1336, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest normalize.normalize_html[8]>", line 1, in <module>
        normalize_html('<a title="bar" HREF="foo">x</a>')
      File "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/normalize.py", line 189, in normalize_html
        parser.feed(chunk.group(0))
      File "/usr/lib64/python3.8/html/parser.py", line 111, in feed
        self.goahead(0)
      File "/usr/lib64/python3.8/html/parser.py", line 171, in goahead
        k = self.parse_starttag(i)
      File "/usr/lib64/python3.8/html/parser.py", line 345, in parse_starttag
        self.handle_starttag(tag, attrs)
      File "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/normalize.py", line 69, in handle_starttag
        self.output += ("=" + '"' + cgi.escape(v,quote=True) + '"')
    AttributeError: module 'cgi' has no attribute 'escape'
**********************************************************************
1 items had failures:
   1 of  10 in normalize.normalize_html
***Test Failed*** 1 failures.
<end of output>
Test time =   0.20 sec
----------------------------------------------------------
Test Failed.
"html_normalization" end time: Jul 18 11:15 BST
"html_normalization" time elapsed: 00:00:00
----------------------------------------------------------

3/9 Testing: spectest_library
3/9 Test: spectest_library
Command: "/usr/bin/python3" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/spec_tests.py" "--no-normalize" "--spec" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/spec.txt" "--library-dir" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir/../src"
Directory: /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir
"spectest_library" start time: Jul 18 11:15 BST
Output:
----------------------------------------------------------
649 passed, 0 failed, 0 errored, 0 skipped
<end of output>
Test time =   0.15 sec
----------------------------------------------------------
Test Passed.
"spectest_library" end time: Jul 18 11:15 BST
"spectest_library" time elapsed: 00:00:00
----------------------------------------------------------

4/9 Testing: pathological_tests_library
4/9 Test: pathological_tests_library
Command: "/usr/bin/python3" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/pathological_tests.py" "--library-dir" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir/../src"
Directory: /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir
"pathological_tests_library" start time: Jul 18 11:15 BST
Output:
----------------------------------------------------------
Testing pathological cases:
nested strong emph [PASSED]
many emph closers with no openers [PASSED]
many emph openers with no closers [PASSED]
many link closers with no openers [PASSED]
many link openers with no closers [PASSED]
mismatched openers and closers [PASSED]
openers and closers multiple of 3 [PASSED]
link openers and emph closers [PASSED]
pattern [ (]( repeated [PASSED]
hard link/emph case [PASSED]
nested brackets [PASSED]
nested block quotes [PASSED]
deeply nested lists [PASSED]
U+0000 in input [PASSED]
backticks [PASSED]
unclosed links A [PASSED]
unclosed links B [PASSED]
0 passed, 0 failed, 0 errored
<end of output>
Test time =   2.78 sec
----------------------------------------------------------
Test Passed.
"pathological_tests_library" end time: Jul 18 11:15 BST
"pathological_tests_library" time elapsed: 00:00:02
----------------------------------------------------------

5/9 Testing: roundtriptest_library
5/9 Test: roundtriptest_library
Command: "/usr/bin/python3" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/roundtrip_tests.py" "--spec" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/spec.txt" "--library-dir" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir/../src"
Directory: /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir
"roundtriptest_library" start time: Jul 18 11:15 BST
Output:
----------------------------------------------------------
Traceback (most recent call last):
  File "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/roundtrip_tests.py", line 45, in <module>
    do_test(converter, test, args.normalize, result_counts)
  File "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/spec_tests.py", line 48, in do_test
    passed = normalize_html(actual_html) == normalize_html(expected_html)
  File "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/normalize.py", line 189, in normalize_html
    parser.feed(chunk.group(0))
  File "/usr/lib64/python3.8/html/parser.py", line 111, in feed
    self.goahead(0)
  File "/usr/lib64/python3.8/html/parser.py", line 171, in goahead
    k = self.parse_starttag(i)
  File "/usr/lib64/python3.8/html/parser.py", line 345, in parse_starttag
    self.handle_starttag(tag, attrs)
  File "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/normalize.py", line 69, in handle_starttag
    self.output += ("=" + '"' + cgi.escape(v,quote=True) + '"')
AttributeError: module 'cgi' has no attribute 'escape'
<end of output>
Test time =   0.16 sec
----------------------------------------------------------
Test Failed.
"roundtriptest_library" end time: Jul 18 11:15 BST
"roundtriptest_library" time elapsed: 00:00:00
----------------------------------------------------------

6/9 Testing: entity_library
6/9 Test: entity_library
Command: "/usr/bin/python3" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/entity_tests.py" "--library-dir" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir/../src"
Directory: /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir
"entity_library" start time: Jul 18 11:15 BST
Output:
----------------------------------------------------------
Testing entities:
2125 passed, 0 failed, 0 errored
<end of output>
Test time =   0.11 sec
----------------------------------------------------------
Test Passed.
"entity_library" end time: Jul 18 11:15 BST
"entity_library" time elapsed: 00:00:00
----------------------------------------------------------

7/9 Testing: spectest_executable
7/9 Test: spectest_executable
Command: "/usr/bin/python3" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/spec_tests.py" "--no-normalize" "--spec" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/spec.txt" "--program" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir/../src/cmark"
Directory: /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir
"spectest_executable" start time: Jul 18 11:15 BST
Output:
----------------------------------------------------------
649 passed, 0 failed, 0 errored, 0 skipped
<end of output>
Test time =   3.28 sec
----------------------------------------------------------
Test Passed.
"spectest_executable" end time: Jul 18 11:15 BST
"spectest_executable" time elapsed: 00:00:03
----------------------------------------------------------

8/9 Testing: smartpuncttest_executable
8/9 Test: smartpuncttest_executable
Command: "/usr/bin/python3" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/spec_tests.py" "--no-normalize" "--spec" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/smart_punct.txt" "--program" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir/../src/cmark --smart"
Directory: /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir
"smartpuncttest_executable" start time: Jul 18 11:15 BST
Output:
----------------------------------------------------------
16 passed, 0 failed, 0 errored, 0 skipped
<end of output>
Test time =   0.24 sec
----------------------------------------------------------
Test Passed.
"smartpuncttest_executable" end time: Jul 18 11:15 BST
"smartpuncttest_executable" time elapsed: 00:00:00
----------------------------------------------------------

9/9 Testing: regressiontest_executable
9/9 Test: regressiontest_executable
Command: "/usr/bin/python3" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/spec_tests.py" "--no-normalize" "--spec" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/test/regression.txt" "--program" "/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir/../src/cmark"
Directory: /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu/testdir
"regressiontest_executable" start time: Jul 18 11:15 BST
Output:
----------------------------------------------------------
12 passed, 0 failed, 0 errored, 0 skipped
<end of output>
Test time =   0.22 sec
----------------------------------------------------------
Test Passed.
"regressiontest_executable" end time: Jul 18 11:15 BST
"regressiontest_executable" time elapsed: 00:00:00
----------------------------------------------------------

End testing: Jul 18 11:15 BST
```

**Comment by nwellnhof** [comment]
Apparently, `cgi.escape` was removed in Python 3.8.

**Comment by jgm** [comment]
Is this not fixed by 68c3a91166347a32a57fb81223750a63cfd92105 ?

**Comment by kloczek** [comment]
Yep it works with that patch
```
+ /usr/bin/make -O -j48 V=1 VERBOSE=1 -C x86_64-redhat-linux-gnu test
make: Entering directory '/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu'
Running tests...
/usr/bin/ctest --force-new-ctest-process
Test project /home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu
    Start 1: api_test
1/9 Test #1: api_test .........................   Passed    0.01 sec
    Start 2: html_normalization
2/9 Test #2: html_normalization ...............   Passed    0.16 sec
    Start 3: spectest_library
3/9 Test #3: spectest_library .................   Passed    0.12 sec
    Start 4: pathological_tests_library
4/9 Test #4: pathological_tests_library .......   Passed    2.83 sec
    Start 5: roundtriptest_library
5/9 Test #5: roundtriptest_library ............   Passed    0.27 sec
    Start 6: entity_library
6/9 Test #6: entity_library ...................   Passed    0.11 sec
    Start 7: spectest_executable
7/9 Test #7: spectest_executable ..............   Passed    3.11 sec
    Start 8: smartpuncttest_executable
8/9 Test #8: smartpuncttest_executable ........   Passed    0.20 sec
    Start 9: regressiontest_executable
9/9 Test #9: regressiontest_executable ........   Passed    0.18 sec

100% tests passed, 0 tests failed out of 9

Total Test time (real) =   7.01 sec
make: Leaving directory '/home/tkloczko/rpmbuild/BUILD/cmark-0.29.0/x86_64-redhat-linux-gnu'
```
Thx

Closing
PS. Are you going to make new release soon?
